pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol

```diff
@@ -59,73 +85,345 @@ contract LiquidityBootstrappingPool is BaseWeightedPool, ReentrancyGuard {
+        uint256 sumWeights = 0;
+
+        for (uint8 i = 0; i < totalTokens; i++) {
+            _require(endWeights[i] >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
+
+            // update the end weights in memory
+            newPoolState = newPoolState.insertUint16(endWeights[i].compress16(), _END_WEIGHT_OFFSET + i * 16);
+
+            sumWeights = sumWeights.add(endWeights[i]);
+        }
+
+        _require(sumWeights == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        _poolState = newPoolState;
+
+        emit GradualWeightUpdateScheduled(effectiveStartTime, endTime, startWeights, endWeights);
+    }
+
+    // Internal functions
+
     function _getNormalizedWeight(IERC20 token) internal view override returns (uint256) {
+        uint8 i;
+
+        // First, convert token address to a token index
+
         // prettier-ignore
-        if (token == _token0) { return _getNormalizedWeight(0); }
-        else if (token == _token1) { return _getNormalizedWeight(1); }
-        else if (token == _token2) { return _getNormalizedWeight(2); }
-        else if (token == _token3) { return _getNormalizedWeight(3); }
+        if (token == _token0) { i = 0; }
+        else if (token == _token1) { i = 1; }
+        else if (token == _token2) { i = 2; }
+        else if (token == _token3) { i = 3; }
         else {
             _revert(Errors.INVALID_TOKEN);
         }
+
+        bytes32 poolState = _poolState;
+
+        return _getNormalizedWeightByIndex(i, poolState);
+    }
+
+    function _getNormalizedWeightByIndex(uint8 i, bytes32 poolState) internal view returns (uint256) {
+        uint256 startWeight = poolState.decodeUint32(i * 32).uncompress32();
+        uint256 endWeight = poolState.decodeUint16(_END_WEIGHT_OFFSET + i * 16).uncompress16();
+
+        // Return 0 - 1 value representing how much of the period has elapsed
+        // Will return 0 if it hasn't started, and 1 if it's past the end
+        uint256 pctProgress = _calculateProgress(poolState);
+
+        return _interpolateWeight(startWeight, endWeight, pctProgress);
     }
 
     function _getNormalizedWeights() internal view override returns (uint256[] memory) {
         uint256 totalTokens = _getTotalTokens();
         uint256[] memory normalizedWeights = new uint256[](totalTokens);
 
+        bytes32 poolState = _poolState;
+
         // prettier-ignore
         {
-            if (totalTokens > 0) { normalizedWeights[0] = _getNormalizedWeight(0); } else { return normalizedWeights; }
-            if (totalTokens > 1) { normalizedWeights[1] = _getNormalizedWeight(1); } else { return normalizedWeights; }
-            if (totalTokens > 2) { normalizedWeights[2] = _getNormalizedWeight(2); } else { return normalizedWeights; }
-            if (totalTokens > 3) { normalizedWeights[3] = _getNormalizedWeight(3); } else { return normalizedWeights; }
+            if (totalTokens > 0) { normalizedWeights[0] = _getNormalizedWeightByIndex(0, poolState);
```

Comment:
Facu used a nice format for this sort of thing in #673. This block would then look like
```
if (totalTokens == 0)  return normalizedWeights;
normalizedWeights[0] = _getNormalizedWeightByIndex(0, poolState);
if (totalTokens == 1  return normalizedWeights;
normalizedWeights[1] = _getNormalizedWeightByIndex(1, poolState);
if (totalTokens == 2)  return normalizedWeights;
normalizedWeights[2] = _getNormalizedWeightByIndex(2, poolState);
if (totalTokens == 3)  return normalizedWeights;
normalizedWeights[3] = _getNormalizedWeightByIndex(3, poolState);
```

We could also remove some of the redundant checks as so:
```
normalizedWeights[0] = _getNormalizedWeightByIndex(0, poolState);
normalizedWeights[1] = _getNormalizedWeightByIndex(1, poolState);
if (totalTokens == 2)  return normalizedWeights;
normalizedWeights[2] = _getNormalizedWeightByIndex(2, poolState);
if (totalTokens == 3)  return normalizedWeights;
normalizedWeights[3] = _getNormalizedWeightByIndex(3, poolState);
```

Reply:
Ok; will update